crates/fluvio/src/consumer/retry.rs

```diff
@@ -0,0 +1,279 @@
+#[cfg(target_arch = "wasm32")]
+type BoxPendingFuture = Pin<
+    Box<
+        dyn Future<
+                Output = Option<
+                    Result<(ConsumerRecord, BoxConsumerStream, Option<i64>), ErrorCode>,
+                >,
+            > + 'static,
+    >,
+>;
+#[cfg(not(target_arch = "wasm32"))]
+type BoxPendingFuture = Pin<
+    Box<
+        dyn Future<
+                Output = Option<
+                    Result<(ConsumerRecord, BoxConsumerStream, Option<i64>), ErrorCode>,
+                >,
+            > + Send
+            + 'static,
+    >,
+>;
+
+#[derive(Clone)]
+pub struct ConsumerWithRetryInner {
+    fluvio_client: Arc<Fluvio>,
+    next_offset_to_read: Option<i64>,
+    consumer_config: ConsumerConfigExt,
+}
+
+/// A consumer stream that automatically retries on failure.
+pub struct ConsumerWithRetry {
+    inner: ConsumerWithRetryInner,
+    current_stream: Option<BoxConsumerStream>,
+    pending: Option<BoxPendingFuture>,
+}
+
+impl Stream for ConsumerWithRetry {
+    type Item = Result<ConsumerRecord, ErrorCode>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let self_mut = self.get_mut();
+
+        // If there is no pending future, create one using the current stream.
+        if self_mut.pending.is_none() {
+            let current_stream = self_mut
+                .current_stream
+                .take()
+                .expect("current_stream must be available");
+            let client = Arc::clone(&self_mut.inner.fluvio_client);
+            self_mut.pending = Some(Box::pin(Self::consumer_with_retry(
+                self_mut.inner.clone(),
+                client,
+                current_stream,
+            )));
+        }
+
+        // Poll the pending future and update the internal state.
+        if let Some(ref mut fut) = self_mut.pending {
+            match fut.as_mut().poll(cx) {
+                Poll::Ready(Some(Ok((record, updated_stream, new_offset)))) => {
+                    self_mut.current_stream = Some(updated_stream);
+                    self_mut.inner.next_offset_to_read = new_offset;
+                    self_mut.pending = None;
+                    Poll::Ready(Some(Ok(record)))
+                }
+                Poll::Ready(Some(Err(e))) => {
+                    self_mut.pending = None;
+                    Poll::Ready(Some(Err(e)))
+                }
+                Poll::Ready(None) => {
+                    self_mut.pending = None;
+                    Poll::Ready(None)
+                }
+                Poll::Pending => Poll::Pending,
+            }
+        } else {
+            Poll::Ready(None)
+        }
+    }
+}
+
+impl ConsumerStream for ConsumerWithRetry {
+    fn offset_commit(&mut self) -> std::result::Result<(), ErrorCode> {
+        if let Some(ref mut stream) = self.current_stream {
+            stream.offset_commit()
+        } else {
+            warn!("Can't commit, no current stream");
```

Comment:
probably this return error since caller should no offset commit should not work.  or this could be made async.   ignoring is not good option

Reply:
Fixed! returning a error now